modules/fedimint-ecash-migration-client/src/lib.rs

```diff
@@ -57,28 +104,969 @@ impl ClientModule for EcashMigrationClientModule {
+
+    #[cfg(feature = "cli")]
+    async fn handle_cli_command(
+        &self,
+        args: &[std::ffi::OsString],
+    ) -> anyhow::Result<serde_json::Value> {
+        cli::handle_cli_command(self, args).await
+    }
+}
+
+/// Metadata stored with the register liability transfer operation
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct RegisterTransferOperationMeta {
+    pub spend_book_merkle_root: MerkleRoot<Nonce>,
+    pub spend_book_entries: u64,
+    pub key_set_hash: KeySetHash,
+    pub out_point: OutPoint,
+}
+
+/// Metadata stored with the fund liability transfer operation
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct FundTransferOperationMeta {
+    pub transfer_id: TransferId,
+    pub amount: Amount,
+    pub change: OutPointRange,
+}
+
+/// Metadata stored with the redeem origin ecash operation
+#[derive(Debug, Clone, Serialize, Deserialize, Encodable, Decodable)]
+pub struct RedeemOriginEcashOperationMeta {
+    pub transfer_id: TransferId,
+    pub amount: Amount,
+    pub change: OutPointRange,
+}
+
+/// Read the origin federation's mint keyset from a config file.
+///
+/// The config file should be in JSON format as produced by
+/// `fedimint-cli --data-dir <data-dir> config` on the origin federation.
+///
+/// # Errors
+/// - If the origin config file cannot be read or parsed.
+pub async fn read_origin_keyset(
+    origin_config_path: impl AsRef<Path>,
+) -> anyhow::Result<Tiered<AggregatePublicKey>> {
+    let origin_config_path = origin_config_path.as_ref();
+
+    let origin_config_json = tokio::fs::read_to_string(origin_config_path)
+        .await
+        .with_context(|| {
+            format!(
+                "Failed to read origin config file: {}",
+                origin_config_path.display()
+            )
+        })?;
+
+    let origin_client_config: JsonClientConfig = serde_json::from_str(&origin_config_json)
+        .with_context(|| {
+            format!(
+                "Failed to parse origin config as ClientConfig: {}",
+                origin_config_path.display()
+            )
+        })?;
+
+    let origin_mint_config_json: JsonWithKind = origin_client_config
+        .modules
+        .values()
+        .find(|module_config| module_config.kind() == &MintCommonInit::KIND)
+        .context("Mint module not found in origin config")?
+        .clone();
+
+    let origin_mint_config: MintClientConfig =
+        serde_json::from_value(origin_mint_config_json.value().clone()).with_context(|| {
+            format!(
+                "Failed to parse origin config as MintClientConfig: {}",
+                origin_config_path.display()
+            )
+        })?;
+
+    Ok(origin_mint_config.tbs_pks)
+}
+
+/// Compute the hash of a tiered keyset.
+pub fn hash_keyset(keyset: &Tiered<AggregatePublicKey>) -> KeySetHash {
+    KeySetHash(keyset.consensus_hash_sha256())
+}
+
+/// Open a spend book file and return a stream of nonces.
```

Comment:
Seems like a very raw ah-hoc format, without versioning or ability to add helpful metadata etc.

Though if we formalize it into something, we probably want to retain ability to stream the output, which is not always easily doable with standard encoding libraries.

We might do some midle-ground, and e.g. make it text file, that looks a bit like http request

```
version: 0
key1: value1
key2: value
---
<lines with nonces>
```

or something. Not a big complication really, but would make it somewhat more flexible. The header part could be even a single line of json.